src/game.rs

```diff
@@ -2,40 +2,99 @@ use crate::cell::Cell;
 use crate::game_error::GameError;
 use num::traits::Num;
 use std::cmp::PartialOrd;
-
+use std::collections::HashSet;
 pub type Position = (usize, usize);
 
 ///Represents an instance of the Game of Life, it's generated with an initial state of living
 ///cells, and it's state can be internally modified to be the next iteration of the game by just calling `next()`
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct Game {
     grid: Vec<Vec<Cell>>,
-    alive_cells: Vec<Position>,
+    alive_cells: HashSet<Position>,
 }
 
 impl Game {
-    /// Arguments:
-    /// - `cells`, a list of initial living cells.
-    /// - `dimensions`, a (usize, usize) which indicates the size of the grid (width, height)
-    ///
-    /// Returns:
-    /// - An instance of Game with size `dimensions` and all cells in `living_cells` alive.
-    /// - An error if `cells` contains out of bound positions with respect to `dimensions`.
-    pub fn from_cells(dimensions: Position, cells: &Vec<Position>) -> Result<Self, GameError> {
-        let (width, height) = dimensions;
-        let mut grid = vec![vec![Cell::new(); width]; height];
-
-        for (cell_i, cell_j) in cells.iter() {
-            //Return error if the cell position is out of bounds
-            if !Self::in_bounds(*cell_i, *cell_j, width, height) {
-                return Err(GameError::OutOfBoundsGridAccess((*cell_i, *cell_j)));
-            }
-            //Otherwise, give life to the cell and return OK
-            grid[*cell_i][*cell_j].give_life();
+    /// Initializes a grid of height dimensions.0 and height dimensions.1
+    pub fn of_size(dimensions: Position) -> Self {
+        let (height, width) = dimensions;
+        Game {
+            grid: vec![vec![Cell::new(); width]; height],
+            alive_cells: HashSet::new(),
+        }
+    }
+
+    /// Initializes a game with set size and already alive cells
+    pub fn from_size_and_cells(
+        dimensions: Position,
+        cells: &[Position],
+    ) -> Result<Self, GameError> {
+        let mut game = Game::of_size(dimensions);
+        game.give_life_list(cells)?;
+        Ok(game)
+    }
+
+    /// Kills the cell if it's alive, gives life to it if it's dead.
+    pub fn toggle_cell(&mut self, pos: (usize, usize)) -> Result<(), GameError> {
+        if self.alive_cells.contains(&pos) {
+            self.kill(pos)
+        } else {
+            self.give_life(pos)
+        }
+    }
+
+    /// Kills a specific cell, returns error when out of bounds
+    pub fn kill(&mut self, pos: Position) -> Result<(), GameError> {
+        self.check_in_bounds(pos)?;
+        let cell = &mut self.grid[pos.0][pos.1];
+        cell.kill();
+        self.alive_cells.remove(&pos);
+        Ok(())
+    }
+
+    /// Gives life to a specific cell, returns error when out of bounds
+    pub fn give_life(&mut self, pos: Position) -> Result<(), GameError> {
```

Comment:
why do you have a Result here?

Reply:
Oh alright for check_in_bounds

Reply:
do you really need to do the bound checking? I mean, if the you are out of bounds you have a bug in your program, but if everything is correct, then why would you want to check that?

Reply:
It really helped me find bugs when i was developing the game loop.

This way you don't allow any of the methods of Game to panic, you leave the responsibility of handling the errors to the caller, in this case the loop in main.rs. If anybody else wanted to implement a game which used this same module, the errors would indicate to them if their implementation has indexing errors.

It's a little bit annoying to read, but I thought it was better this way, i might be wrong tho. I also just wanted to practice error definitions and handling